org.lflang/src/lib/core/federated/RTI/rti.c

```diff
@@ -689,7 +693,28 @@ bool send_tag_advance_if_appropriate(federate_t* fed) {
 }
 
 /**
- * Handle a logical tag complete (LTC) message.
+ * For all federates transitively downstream of the specified federate, determine
+ * whether they should be sent a TAG or PTAG and send it if so.
+ *
+ * This assumes the caller holds the mutex.
+ *
+ * @param fed The upstream federate.
+ * @param visited An array of booleans used to determine whether a federate has
+ *  been visited (initially all false).
+ */
+void send_TAG_if_transitively_safe_to_all_downstream(federate_t* fed, bool visited[]) {
+	visited[fed->id] = true;
+	for (int i = 0; i < fed->num_downstream; i++) {
+		federate_t* downstream = &_RTI.federates[fed->downstream[i]];
+		if (visited[downstream->id]) continue;
+		send_TAG_if_transitively_safe(downstream);
+		send_TAG_if_transitively_safe_to_all_downstream(downstream, visited);
```

Comment:
I think the names of these functions can be improved. I would call one `grant_advancement` (because it could also send `PTAG`, not just `TAG`) and the other `traverse_and_grant_advancement`. I don't think it is necessary to explain in the name the fact notices will only be sent granted that certain conditions are true. Comments are adequate for that.

Reply:
`grant_advancement` can be misleading since these functions sometimes don't issue a `TAG` or `PTAG`. Also advancement might be a bit vague. Does it mean tag advancement in all contexts?

Reply:
How about `grant_TAG_if_safe` and `grant_TAG_downstream_if_safe`?

Reply:
It chose the term "advancement" because it generalizes the different kinds of grants that are issued (PTAG/TAG); stating that it only grants a TAG is inaccurate. The `if_safe` part strikes me as unnecessary. The function `malloc` also doesn't always allocate memory, yet we don't call it `malloc_if_memory_available`. That's just too verbose. Finally, the `downstream` aspect seems to convey another implementation detail that is not relevant to the caller of the API who just wants to traverse the graph of federates and grant advancement.

Reply:
In the case of `malloc`, it doesn't allocate memory if an error occurs. In this function, not sending a TAG/PTAG is not an error. This conditional sending of TAG/PTAG is part of the main logic of this function. I think it should be clear that calling this function will not necessarily yield a TAG/PTAG.

In terms of advancement, as I said, I think it is vague. What does it mean to grant an advancement? These functions don't even necessarily grant a tag advancement even if they send a TAG. If a PTAG has been sent to a federate, and then later this function sends a TAG, it will just unblock control reactions.

Reply:
Maybe `grant_advance_grant_if_safe`? I think that downstream is pretty important. Traverse doesn't indicate the direction of the traversal. Calling the second function will send TAG/PTAG downstream only if it is safe.

Reply:
I'm fine with `send_advance _grant` and `send_advance_grants_downstream`, but I haven't changed my opinion on the `if_safe` part. IMO, to pursue safety is the default in programming; only doing something distinctly unsafe requires raising extra awareness to where I would factor it into the name of a function. I think this boils down to a difference in API design aesthetics that's not worth debating...

Reply:
So maybe going back to `_if_apropriate` is a better choice than `_is_safe`.

I think that it should be clear that this function is not going to send a TAG/PTAG just because it is called. It should be called later if the circumstances change.

Reply:
I don't think `if_appropriate` is better. I also don't think there's anything that I can add to this discussion at this point, and it doesn't seem like we're able to agree on things, so perhaps someone else can comment on this. We can also declare this a non-issue and move on.

Reply:
But there is a logistical issue as well. We'll end up with the following four functions:
```
send_advance_grant
send_advance_grant_downstream
send_tag_advance_grant  // Sends TAG to a federate without any calculations
send_provisional_tag_advance_grant // Sends PTAG to a federate without any calculations
```
If you give me a concrete name for all of these, I will go ahead and rename them to your suggestion.

Reply:
I think I'm with Soroush on this, though honestly I'm not completely sure. If you have a method called "send_advance_grant" and it doesn't send an advance grant, you would have to assume it is defective, right?  What this method does, I think, is to determine whether it is "safe" (in the "safe to process" sense) to send an advance grant, and, if so, send it.  So to me, send_advance_grant_if_safe seems like a pretty descriptive name.

Reply:
In any case, I think that `send_advance_grant_if_safe` is an improvement over the current name, so let's go with that then. To be consistent, I suppose the recursive function should be called `send_downstream_advance_grants_if_safe`? IDK, I don't like these super long function names, but again, I think it's better than `send_TAG_if_transitively_safe_to_all_downstream`.

Reply:
Instead of an `if_safe` suffix, a `try` prefix could also be used to convey the tentative nature of the operation. I think that's more common.

Reply:
"try" implies that there may be some reason outside the control of the system that it would fail, like network connectivity.

A meta issue: In an OO language, this function would be private.  It is designed for internal use. I personally prefer long descriptive names for such functions because it reduces fatigue reading the code. I spend less time jumping to the function to find its documentation.